packages/next/lib/check-custom-routes.ts

```diff
@@ -0,0 +1,88 @@
+export type Rewrite = {
+  source: string
+  destination: string
+}
+
+export type Redirect = Rewrite & {
+  statusCode?: number
+}
+
+const allowedStatusCodes = new Set([301, 302, 303, 307, 308])
+
+export default function checkCustomRoutes(
+  routes: Array<Rewrite | Redirect>,
+  type: 'redirect' | 'rewrite'
+): void {
+  let numInvalidRoutes = 0
+  let hadInvalidStatus = false
+  const isRedirect = type === 'redirect'
+  const allowedKeys = new Set([
+    'source',
+    'destination',
+    ...(isRedirect ? ['statusCode'] : []),
+  ])
+
+  for (const route of routes) {
+    const keys = Object.keys(route)
+    const invalidKeys = keys.filter(key => !allowedKeys.has(key))
+    const invalidParts = []
+
+    // TODO: investigate allowing RegExp directly
+    if (!route.source) {
+      invalidParts.push('`source` is missing')
+    } else if (typeof route.source !== 'string') {
+      invalidParts.push('`source` is not a string')
+    } else if (!route.source.startsWith('/')) {
+      invalidParts.push('`source` does not start with /')
+    }
+
+    if (!route.destination) {
+      invalidParts.push('`destination` is missing')
+    } else if (typeof route.destination !== 'string') {
+      invalidParts.push('`destination` is not a string')
+    } else if (!route.destination.startsWith('/')) {
+      invalidParts.push('`destination` does not start with /')
+    }
+
+    if (isRedirect) {
+      const redirRoute = route as Redirect
+
+      if (
+        redirRoute.statusCode &&
+        !allowedStatusCodes.has(redirRoute.statusCode)
+      ) {
+        hadInvalidStatus = true
+        invalidParts.push(`\`statusCode\` is not undefined or valid statusCode`)
```

Comment:
This message is confusing, we should mention all the `allowedStatusCodes` otherwise it'll be impossible to guess what works when putting in a wrong value.

Reply:
It logs them on the bottom of the output if they had any invalid status codes so we don't repeat them each time

Reply:
I mean, it's not clear what a *valid* statusCode is.

Reply:
Do you think the wording of the valid statusCodes message at the bottom needs to be more verbose?

<img width="470" alt="68998010-585ceb80-0872-11ea-8215-468b9ad3eed1" src="https://user-images.githubusercontent.com/22380829/69648063-37458900-1030-11ea-8f85-006ea2d1dadf.png">

Reply:
Ah I see 👍